octave: Do not --enable-link-all-dependencies on Darwin

ravenjoad opened the pull request:
Some Octave packages ship C/C++/Fortran code that needs to be compiled and linked against Octave's internal shared objects. Previously, `ld` could not find these packages because `$LD_LIBRARY_PATH` was empty.

Closes #368187.

`nixpkgs-review` marked some new packages as broken (bim, fits, msh), but ltfat builds correctly.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

paparodeo commented on pkgs/development/interpreters/octave/build-octave-package.nix:
```diff
@@ -123,6 +123,10 @@ stdenv.mkDerivation (
+    # Add Octave's shared objects to ld library path for some packages which
+    # need to link their C/C++/Fortran code against Octave's internals.
+    LD_LIBRARY_PATH = lib.makeLibraryPath nativeBuildInputs';
+
```
I'm a little confused about how `LD_LIBRARY_PATH` helps.

however, was able to get the build to work with this
```diff
diff --git a/pkgs/development/octave-modules/ltfat/default.nix b/pkgs/development/octave-modules/ltfat/default.nix
index 43fe020b322a..d22a78938b5a 100644
--- a/pkgs/development/octave-modules/ltfat/default.nix
+++ b/pkgs/development/octave-modules/ltfat/default.nix
@@ -9,6 +9,7 @@
   lapack,
   blas,
   portaudio,
+  octave,
   jdk,
 }:
 
@@ -21,6 +22,8 @@ buildOctavePackage rec {
     sha256 = "sha256-FMDZ8XFhLG7KDoUjtXvafekg6tSltwBaO0+//jMzJj4=";
   };
 
+  NIX_LDFLAGS = "-L${lib.getLib octave}/lib/octave/9.3.0";
+
   buildInputs = [
     fftw
     fftwSinglePrec
```

on darwin and linux tho didn't test the plugin to see if it actually works -- there were some linker warnings on darwin.

paparodeo replied:
ok looking at the bundle in `oct/Makefile_unix` it seems to export `LDFLAGS`

```
export LDFLAGS := -L$(shell $(MKOCTFILE) -p LIBDIR) $(FLIBS) $(LAPACK_LIBS) $(BLAS_LIBS) $(FFTW_LIBS)
```
and thru some experimentation when `LDFLAGS` are set I guess that overrides the flags used by `mkoctfile` because
```
make -f Makefile_unix comp_atrousfilterbank_td.oct
mkoctfile -strip -I../src/modules/libltfat/include -DNDEBUG -L../lib -lltfat comp_atrousfilterbank_td.cc
mkoctfile: stripping disabled on this platform
ld: library not found for -loctinterp
```
but just can comment out `LDFLAGS` in the makefile and then the command succeeds.

so this seems like an upstream issue.

paparodeo replied:
ok -- apologies for all the posts. this seems like a workaround until the problem is fixed in the upstream makefile.

```diff
diff --git a/pkgs/development/octave-modules/ltfat/default.nix b/pkgs/development/octave-modules/ltfat/default.nix
index 43fe020b322a..d98253467190 100644
--- a/pkgs/development/octave-modules/ltfat/default.nix
+++ b/pkgs/development/octave-modules/ltfat/default.nix
@@ -1,4 +1,5 @@
 {
+  stdenv,
   buildOctavePackage,
   lib,
   fetchurl,
@@ -32,6 +33,12 @@ buildOctavePackage rec {
     jdk
   ];
 
+  # work around issue in upstream oct/Makefile_unix where it doesn't add
+  # -L$(mkoctfile -p OCTLIBDIR) to exported LDFLAGS
+  postPatch = lib.optionalString stdenv.hostPlatform.isDarwin ''
+    NIX_LDFLAGS+=" -L$(mkoctfile -p OCTLIBDIR)"
+  '';
+
   meta = with lib; {
     name = "The Large Time-Frequency Analysis Toolbox";
     homepage = "https://octave.sourceforge.io/ltfat/index.html";

```

paparodeo replied:
ok -- sorry for all the posts. I just realized that can use `patches` in the octave builder. `LDFLAGS` in the env will clobber `mkoctfile` ldflags, so make sure to call `mkoctfile -p LDFLAGS` first.

should file an issue: https://github.com/ltfat/ltfat/issues

<details><summary>don't clobber ldflags.diff</summary>

```diff
diff --git a/pkgs/development/octave-modules/ltfat/append-ldflags.diff b/pkgs/development/octave-modules/ltfat/append-ldflags.diff
new file mode 100644
index 000000000000..f0f15bc1b5fe
--- /dev/null
+++ b/pkgs/development/octave-modules/ltfat/append-ldflags.diff
@@ -0,0 +1,13 @@
+diff --git a/oct/Makefile_unix b/oct/Makefile_unix
+index 21e026b..c9d767a 100644
+--- a/oct/Makefile_unix
++++ b/oct/Makefile_unix
+@@ -36,7 +36,7 @@ endif
+ export CXXFLAGS := $(shell $(MKOCTFILE) -p CXXFLAGS) -std=gnu++11 -Wall -DLTFAT_LARGEARRAYS $(OPTCXXFLAGS)
+ # export is necessary, otherwise LFLAGS won't have any effect
+ # at least on Windows and on Mac
+-export LDFLAGS := -L$(shell $(MKOCTFILE) -p LIBDIR) $(FLIBS) $(LAPACK_LIBS) $(BLAS_LIBS) $(FFTW_LIBS) 
++export LDFLAGS := $(shell $(MKOCTFILE) -p LDFLAGS) $(FLIBS) $(LAPACK_LIBS) $(BLAS_LIBS) $(FFTW_LIBS)
+ undefine LFLAGS
+ unexport LFLAGS
+ 
diff --git a/pkgs/development/octave-modules/ltfat/default.nix b/pkgs/development/octave-modules/ltfat/default.nix
index 43fe020b322a..c5462f6392ea 100644
--- a/pkgs/development/octave-modules/ltfat/default.nix
+++ b/pkgs/development/octave-modules/ltfat/default.nix
@@ -32,6 +32,11 @@ buildOctavePackage rec {
     jdk
   ];
 
+  patches = [
+    # don't clobber mkoctfile ldflags, append them
+    ./append-ldflags.diff
+  ];
+
   meta = with lib; {
     name = "The Large Time-Frequency Analysis Toolbox";
     homepage = "https://octave.sourceforge.io/ltfat/index.html";

```
</details>

ravenjoad committed:
octave: Do not --enable-link-all-dependencies on Darwin

It seems like upstream Octave removed the requirement that Darwin
default to linking all of its dependencies[1]. Some packages (ltfat)
cannot link against Octave for some reason. Removing this optinal
configure flag fixes this issue.

[1] https://github.com/gnu-octave/octave/commit/d4479bd8aef35911e07851ef3aee89ef3954604b

paparodeo reviewed (commented):
this change looks good but the title and commit message need to match -- perhaps:
```
octave: remove --enable-link-all-dependencies

darwin set --enable-link-all-dependencies which was the default for
homebrew and octave for a little bit but both reverted the change a
a couple of years ago:

https://github.com/Homebrew/homebrew-core/commit/a01651b19e4babace008b7dd94899807f1818116
https://github.com/gnu-octave/octave/commit/d4479bd8aef35911e07851ef3aee89ef3954604b
```
or something.

note: https://github.com/NixOS/nixpkgs/issues/368187#issuecomment-2563601215 has a bunch of references to when `--enable-link-all-dependencies` was removed in home-brew and disabled by default on octave.

I built this on aarch64-darwin and enabled tests and results were identical to that of x64 linux.

[edit] NOTE: I'm building on darwin with this change https://github.com/NixOS/nixpkgs/pull/367594 which is required since https://github.com/NixOS/nixpkgs/pull/357259 broke arpack on darwin

KarlJoad commented:
@paparodeo, excellent! 

Yep! I'll update the necessary PR stuff soon (and rebase the `LD_LIBRARY_PATH` out of existence).

KarlJoad commented:
Given that @doronbehar is the current maintainer of the Octave definition, I also want to see his response to this as well.

doronbehar commented on pkgs/development/interpreters/octave/build-octave-package.nix:
```diff
@@ -123,10 +123,6 @@ stdenv.mkDerivation (
 
-    # Add Octave's shared objects to ld library path for some packages which
-    # need to link their C/C++/Fortran code against Octave's internals.
-    LD_LIBRARY_PATH = lib.makeLibraryPath nativeBuildInputs';
```
I'm confused. As a start can we squash this addition and then removal and hence the whole PR to 1 commit?

KarlJoad replied:
Yes, that was the plan. The `LD_LIBRARY_PATH` line commit will be removed.

doronbehar reviewed (commented):
IIRC the ltfat package was broken on all platforms only when you actually try to run a `octave.withPackages (ps: [ ps.ltfat ])`

doronbehar commented:
> IIRC the ltfat package was broken on all platforms only when you actually try to run a `octave.withPackages (ps: [ ps.ltfat ])`

I didn't remember correctly, and the issue I was thinking about was [the issue fixed by this PR](https://github.com/NixOS/nixpkgs/issues/368187), but rather:

- https://github.com/NixOS/nixpkgs/issues/328389

doronbehar reviewed (approved):
Changes and commit message LGTM.

7c6f434c reviewed (approved):
Looks reasonable, but I have no clue about Darwin
